datafusion/src/physical_plan/sorts/in_mem_sort.rs

```diff
@@ -0,0 +1,241 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::collections::BinaryHeap;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use arrow::{
+    array::{make_array as make_arrow_array, MutableArrayData},
+    compute::SortOptions,
+    datatypes::SchemaRef,
+    error::{ArrowError, Result as ArrowResult},
+    record_batch::RecordBatch,
+};
+use futures::Stream;
+
+use crate::error::Result;
+use crate::physical_plan::metrics::BaselineMetrics;
+use crate::physical_plan::sorts::{RowIndex, SortKeyCursor};
+use crate::physical_plan::{
+    expressions::PhysicalSortExpr, PhysicalExpr, RecordBatchStream,
+};
+
+/// Merge buffered, self-sorted record batches to get an order.
+///
+/// Internally, it uses MinHeap to reduce extra memory consumption
+/// by not concatenating all batches into one and sorting it as done by `SortExec`.
```

Comment:
I need to study the connection between `SortExec`, `SortPreservingMergeStream` and `InMemSortStream` some more to fully get understand this. I can't help by think that `InMemSortStream` is doing the same thing as `SortPreservingMergeStream` -- and I wonder if we can reuse that same code

Reply:
The main difference between `InMemSortStream` and `SortPreservingMergeStream` lies in assuming different numbers of "entity" (batches for IMSS and streams for SPMS) merged. 

Since `InMemSort` means to merge much more partially ordered "entities", the sorter should reduce the num of comparison for each item pop-up. Hence a MinHeap was introduced.

On the other hand, `InMemSort` is more specialized to have each "entity" only one record batch, therefore simplified logic compared to consider stream continuation in `SortPreservingMergeStream`.

Currently, the common parts `SortKeyCursor` and `RowIndex` for both sorts are extracted to `sorts/mod.rs` reduce duplication.